corehq/ex-submodules/couchforms/models.py

```diff
@@ -346,27 +346,56 @@ def _meta_to_json(meta):
-        self.history.append(XFormOperation(
-            user=user_id,
-            operation='archive',
-        ))
-        self.save()
-        xform_archived.send(sender="couchforms", xform=self)
+    def archive(self, user_id=None, retry_archive=False):
+        if not retry_archive:
+            if self.is_archived:
+                return
+            # If this archive was initiated by a user, delete all other stubs for this action so that this action
+            # isn't overridden
+            from couchforms.models import UnfinishedArchiveStub
+            UnfinishedArchiveStub.objects.filter(user_id=user_id).all().delete()
+        from corehq.form_processor.submission_process_tracker import unfinished_archive
+        with unfinished_archive(instance=self, user_id=user_id, archive=True) as archive_stub:
+            self.doc_type = "XFormArchived"
+
+            self.history.append(XFormOperation(
+                user=user_id,
+                operation='archive',
+            ))
+            archive_stub.archive_history_updated()
+            self.save()
+            xform_archived.send(sender="couchforms", xform=self)
+
+    def unarchive(self, user_id=None, retry_archive=False):
+        if not retry_archive:
+            if not self.is_archived:
+                return
+            # If this unarchive was initiated by a user, delete all other stubs for this action so that this action
+            # isn't overridden
+            from couchforms.models import UnfinishedArchiveStub
+            UnfinishedArchiveStub.objects.filter(user_id=user_id).all().delete()
+        from corehq.form_processor.submission_process_tracker import unfinished_archive
+        with unfinished_archive(instance=self, user_id=user_id, archive=False) as archive_stub:
+            self.doc_type = "XFormInstance"
+            self.history.append(XFormOperation(
+                user=user_id,
+                operation='unarchive',
+            ))
+            archive_stub.archive_history_updated()
+            XFormInstance.save(self)  # subclasses explicitly set the doc type so force regular save
+            xform_unarchived.send(sender="couchforms", xform=self)
+
+    def send_archive_to_kafka(self, user_id):
+        from corehq.form_processor.submission_process_tracker import unfinished_archive
+        with unfinished_archive(instance=self, user_id=user_id, archive=True):
+            self.save()
+            xform_archived.send(sender="couchforms", xform=self)
 
-    def unarchive(self, user_id=None):
-        if not self.is_archived:
-            return
-        self.doc_type = "XFormInstance"
-        self.history.append(XFormOperation(
-            user=user_id,
-            operation='unarchive',
-        ))
-        XFormInstance.save(self)  # subclasses explicitly set the doc type so force regular save
-        xform_unarchived.send(sender="couchforms", xform=self)
+    def send_unarchive_to_kafka(self, user_id):
```

Comment:
can just call `xform_unarchived.send(sender="couchforms", xform=self)` directly from the task